packages/devtools_app_shared/lib/src/service/service_manager.dart

```diff
@@ -521,30 +522,81 @@ class ServiceManager<T extends VmService> {
     // VM service connection spawned from `dart test` or `flutter test`).
     if (packageRootUriString?.endsWith('.dart') ?? false) {
       final rootLibrary = await _mainIsolateRootLibrary();
-      final testTargetFileUriString =
-          (rootLibrary?.dependencies ?? <LibraryDependency>[])
-              .firstWhereOrNull((dep) => dep.prefix == 'test')
-              ?.target
-              ?.uri;
-      if (testTargetFileUriString != null) {
+      if (rootLibrary != null) {
+        packageRootUriString =
+            (await _lookupPackageConfigByEval(rootLibrary)) ??
+                // TODO(kenz): remove this fallback once all test bootstrap
+                // generators include the `packageConfigLocation` constant we
+                // can evaluate.
+                await _lookupTestLibraryByPrefix(rootLibrary, dtdManager);
+      }
+    }
+    _log.fine(
+      '[connectedAppPackageRoot] package root for test target: '
+      '$packageRootUriString',
+    );
+    return packageRootUriString == null
+        ? null
+        : Uri.parse(packageRootUriString);
+  }
+
+  Future<String?> _lookupPackageConfigByEval(Library rootLibrary) async {
+    final eval = EvalOnDartLibrary(
+      rootLibrary.uri!,
+      this.service! as VmService,
+      serviceManager: this,
+      // Swallow exceptions since this evaluation may be called on an older
+      // version of package:test where we do not expect the evaluation to
+      // succeed.
+      logExceptions: false,
+    );
+    final evalDisposable = Disposable();
+    try {
+      final packageConfig = (await eval.evalInstance(
+        'packageConfigLocationn',
+        isAlive: evalDisposable,
+      ))
+          .valueAsString;
+
+      if (packageConfig?.endsWith(packageConfigIdentifier) ?? false) {
         _log.fine(
-          '[connectedAppPackageRoot] detected test library from root library '
-          'imports: $testTargetFileUriString',
+          '[connectedAppPackageRoot] detected test package config from root '
+          'library eval: $packageConfig.',
         );
-        packageRootUriString = await packageRootFromFileUriString(
-          testTargetFileUriString,
-          dtd: dtdManager.connection.value,
-        );
-        _log.fine(
-          '[connectedAppPackageRoot] package root for test target: '
-          '$packageRootUriString',
+        return packageConfig!.substring(
```

Comment:
It looks like this is actually trying to look up the root of the current package, and not the package config URI? The name of the function should be changed then - and if this is actually what you need then we should just write this into the variable instead of the package config location?

In the workspace case, this is going to give the _repo_ root and not the current package root.

Reply:
I have renamed the helper methods and filed https://github.com/flutter/devtools/issues/7944 to track changing the API to take the full path to the package config instead of rebuilding it from the root.